Add child elements category to models tree

JonasDov opened the pull request:
closes #1100 
This change, allows models tree to show category nodes of child elements, when their category differs from parent node's category.

PS. Don't review yet, visibility changes are still needed

grigasp commented on packages/itwin/tree-widget/src/test/trees/models-tree/internal/ModelsTreeVisibilityHandler.test.ts:
```diff
@@ -2535,6 +2535,7 @@ describe("HierarchyBasedVisibilityHandler", () => {
               // FIXME: This is strange from the UX perspective
               groupingNode: ({ elementIds }) => (elementIds.includes(parentElementId) ? "visible" : "hidden"),
               element: ({ elementId }) => (elementId === parentElementId ? "visible" : "hidden"),
+              category: ({ categoryId }) => (categoryId === childCategoryId ? "hidden" : "visible"),
```
flipped the comparison to stay consistent with other checks

```suggestion
              category: ({ categoryId }) => (categoryId === parentCategoryId ? "visible" : "hidden"),
```

grigasp reviewed (commented):
> P.S. Don't review yet, as filtering doesn't work.

Please add tests for filtering

JonasDov commented:
> > P.S. Don't review yet, as filtering doesn't work.
> 
> Please add tests for filtering

Added one tests, and modified existing test a little bit.

grigasp commented on packages/itwin/tree-widget/src/e2e-tests/ModelsTree.test.ts-snapshots/Models-tree-Density-default-instances-focus-1-chromium-linux.png:
I think we should move the mouse pointer away from nodes to avoid the hover effect.

grigasp commented on packages/itwin/tree-widget/src/e2e-tests/ModelsTree.test.ts-snapshots/Models-tree-Density-default-instances-focus-1-chromium-linux.png:
Do you know why "Tag-Category" is in partial display state?

grigasp commented on packages/itwin/tree-widget/src/e2e-tests/ModelsTree.test.ts-snapshots/Models-tree-Density-default-search-1-chromium-linux.png:
Do you know why display state of "Tag-Category" is "off"?

grigasp commented on packages/itwin/tree-widget/src/e2e-tests/ModelsTree.test.ts-snapshots/Models-tree-Density-default-search-1-chromium-linux.png:
Do you know why display state of "E-101 [4-1F4]" is "partial"? Also, shouldn't its class grouping node also be in partial display state?

grigasp reviewed (commented):
There seem to be some issues with display states. 

I understand you're going to update display states handling in a separate PR, and that's fine. But the screenshots show that display states of nodes **above** the new category nodes have changed. I don't understand why.

grigasp commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/models-tree/internal/AlwaysAndNeverDrawnElementInfo.ts:
```diff
@@ -132,66 +152,54 @@ export class AlwaysAndNeverDrawnElementInfo implements IDisposable {
-        ctes: [
-          `
-        ElementInfo(elementId, modelId, categoryId, parentId) AS (
+        ecsql: `
```
why don't we need to select parents of these elements anymore?

grigasp commented on packages/itwin/tree-widget/src/test/trees/models-tree/internal/AlwaysAndNeverDrawnElementInfo.test.ts:
```diff
@@ -37,20 +34,19 @@ describe("AlwaysAndNeverDrawnElementInfo", () => {
-      await using(new AlwaysAndNeverDrawnElementInfo(vp), async (_) => {
+      (() => {
+        using _ = new AlwaysAndNeverDrawnElementInfo(vp);
         expect(event.numberOfListeners).to.eq(1);
-      });
-      await sinon.clock.runAllAsync();
+      })();
```
```suggestion
      {
        using _ = new AlwaysAndNeverDrawnElementInfo(vp);
        expect(event.numberOfListeners).to.eq(1);
      }
```

grigasp commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/models-tree/internal/ModelsTreeIdsCache.ts:
```diff
@@ -376,74 +355,194 @@ export class ModelsTreeIdsCache {
-            CategoryElements(id, modelId, categoryId) AS (
-              SELECT ECInstanceId, Model.Id, Category.Id
+          `
+          ElementsChildrenInfo(Id, CategoryId, ParentId, RootElementCategoryId, ParentHasParent) AS (
```
This CTE selects all elements of given model - that could be millions of elements. Did you check how this query performs in large datasets?

grigasp commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/models-tree/internal/ModelsTreeIdsCache.ts:
```diff
@@ -376,74 +355,194 @@ export class ModelsTreeIdsCache {
+            this.Id id,
+            this.RootElementCategoryId rootElementCategoryId,
+            IFNULL((
+              SELECT 1
+              FROM (
+                SELECT ParentId FROM ElementsChildrenInfo
+              )
+              WHERE ParentId = this.Id
+              LIMIT 1
```
why not just `SELECT 1 FROM ${this._hierarchyConfig.elementClassSpecification} WHERE Parent.Id = this.Id`?

grigasp commented on packages/itwin/tree-widget/src/tree-widget-react/components/trees/models-tree/internal/ModelsTreeIdsCache.ts:
```diff
@@ -376,74 +355,194 @@ export class ModelsTreeIdsCache {
+  /**
+   * Returns elements that are display under the category in models tree.
+   *
+   * There can be two different cases:
+   * 1. If category is directly under model, then it returns only those elements that have children.
+   * 2. If category is under element, then it returns all child elements.
```
why not have two different functions for this?

grigasp commented on packages/itwin/tree-widget/api/tree-widget-react.api.md:
```diff
@@ -90,7 +89,7 @@ interface ChangeGeometricElementsDisplayStateProps {
     categoryId: Id64String;
     // (undocumented)
-    elementIds: Id64Set;
+    elementIds: Map<Id64String, boolean>;
```
breaking change

grigasp commented on packages/itwin/tree-widget/api/tree-widget-react.api.md:
```diff
@@ -179,6 +178,8 @@ interface GetCategoryVisibilityStatusProps {
     // (undocumented)
     modelId: Id64String;
+    // (undocumented)
+    parentElementIds: Id64Array;
```
breaking change

grigasp reviewed (commented):
I think this needs to be reviewed from the perspective of the executed queries.

E.g. when I open http://localhost:3000/?iTwinId=892aa2c9-5be8-4865-9f37-7d4c7e75ebbf&iModelId=77d710e9-83aa-4864-9248-23dfe4bc0047, I see 73 `queryRows` requests. Quite a few of them take around a second. One of them takes almost 2 second and returns ids of 82k elements.
